src/runtime/cuda/nccl_api.cu

```diff
@@ -0,0 +1,832 @@
+  // we assume a poorly partitioned graph, and that there exists the
+  // possibility that each processor could request data from this one.
+
+  // the buffer for us to re-order our requests in
+  IdType * send_idx = static_cast<IdType*>(device->AllocWorkspace(ctx,
+      num_in*sizeof(IdType)));
+
+  std::pair<IdArray, NDArray> part_perm = part->GeneratePermutation(req_idx);
+  const IdType * const perm = static_cast<const IdType*>(part_perm.first->data);
+  const int64_t * const send_sum =
+      static_cast<const int64_t*>(part_perm.second->data);
+
+  // permute requests
+  {
+    const dim3 block(256);
+    const dim3 grid((num_in+block.x-1)/block.x);
+
+    aten::impl::IndexSelectSingleKernel<<<grid, block, 0, stream>>>(
+        static_cast<const IdType*>(req_idx->data),
+        perm,
+        num_in,
+        send_idx);
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  // compute the prefix sum of the indexes this process is requesting
+  int64_t * request_prefix = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  {
+    size_t prefix_workspace_size;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(nullptr, prefix_workspace_size,
+        send_sum, request_prefix, comm_size+1, stream));
+
+    void * prefix_workspace = device->AllocWorkspace(
+        ctx, prefix_workspace_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(prefix_workspace, prefix_workspace_size,
+        send_sum, request_prefix, comm_size+1, stream));
+    device->FreeWorkspace(ctx, prefix_workspace);
+  }
+
+  std::vector<int64_t> request_prefix_host(comm_size+1);
+  device->CopyDataFromTo(
+      request_prefix,
+      0,
+      request_prefix_host.data(),
+      0,
+      request_prefix_host.size()*sizeof(*request_prefix),
+      ctx,
+      DGLContext{kDLCPU, 0},
+      DGLType{kDLInt, sizeof(*request_prefix)*8, 1},
+      stream);
+  device->FreeWorkspace(ctx, request_prefix);
+  CHECK_EQ(request_prefix_host.back(), num_in);
+
+  // communicate the amount requested
+  int64_t * recv_sum = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  comm->AllToAll(send_sum, recv_sum, 1, stream);
+
+  // compute the prefix sum of the requested indexes
+  int64_t * response_prefix = static_cast<int64_t*>(
+      device->AllocWorkspace(ctx, sizeof(int64_t)*(comm_size+1)));
+  {
+    size_t prefix_workspace_size;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(nullptr, prefix_workspace_size,
+        recv_sum, response_prefix, comm_size+1, stream));
+
+    void * prefix_workspace = device->AllocWorkspace(
+        ctx, prefix_workspace_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(prefix_workspace, prefix_workspace_size,
+        recv_sum, response_prefix, comm_size+1, stream));
+    device->FreeWorkspace(ctx, prefix_workspace);
+  }
+  device->FreeWorkspace(ctx, recv_sum);
+
+  // finally copy the prefixsum sum down to the host
+  std::vector<int64_t> response_prefix_host(comm_size+1);
```

Comment:
Do we need to copy prefix sum to CPU?

Reply:
This is needed for the call to NCCL, to specify which portions of the array need to be sent to which processors. See https://github.com/dmlc/dgl/pull/2825/files#diff-5d6e33a4f7ce35295db947aff3e3cb6462f630d6a7b85629c3b2a3eb5c32e98dR82